text/0041-fixed-point.md

```diff
@@ -0,0 +1,177 @@
+
+## Motivation
+[motivation]: #motivation
+
+Fractional values in hardware are usually represented as some form of fixed point value.
+Without a first class fixed point type, the user has to manually manage how the value needs to be shifted to be represented as an integer and keep track of how that interacts with arithmetic operations.
+
+A fixed point type would encode and keep track of the precision through arithmetic operations, as well as provide standard operations for converting values to and from fixed point representations.
+
+## Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+TODO
+
+## Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+This RFC proposes a library addition `amaranth.lib.fixed` with the following contents:
+
+### `fixed.Shape`
+
+`fixed.Shape` is a `ShapeCastable` subclass.
+The following operations are defined on it:
+
+- `fixed.Shape(shape, f_bits)`: Create a `fixed.Shape` with underlying storage `shape` and `f_bits` fractional bits.
+- The signedness is inherited from `shape`, so a `fixed.Shape(signed(16), 12)` would be a signed fixed-point number, 16 bits wide with 12 fractional bits.
+- A `fixed.Shape` may be constructed using the following aliases:
+    - `SQ(i_bits, f_bits)` is an alias for `fixed.Shape(signed(i_bits + f_bits), f_bits)`.
+    - `UQ(i_bits, f_bits)` is an alias for `fixed.Shape(unsigned(i_bits + f_bits), f_bits)`.
+- `.i_bits`, `.f_bits`, `.signed`: Width and signedness properties of the `fixed.Shape`.
+    - `.i_bits` includes the sign bit. That is, for `fixed.Shape(signed(16), 12)`, `.i_bits == 4`.
+- `.const(value)`: Create a `fixed.Const` from `value`.
+- `.as_shape()`: Return the underlying `Shape`.
+- `.__call__(target)`: Create a `fixed.Value` over `target`.
+- `min()`, `max()`: Returns a `fixed.Const` representing the minimum and maximum representable values of a shape.
+
+### `fixed.Value`
+
+`fixed.Value` is a `ValueCastable` subclass.
+The following operations are defined on it:
+
+- `fixed.Value(shape, target)`: Create a `fixed.Value` with `shape` over `target`.
+- `fixed.Value.cast(value, f_bits=0)`: Cast `value` to a `fixed.Value`.
+- `.i_bits`, `.f_bits`, `.signed`: Width and signedness properties.
+- `.shape()`: Return the `fixed.Shape` this was created from.
+- `.as_value()`: Return the underlying value.
+- `.numerator()`: Return `as_value()` cast to the appropriate signedness.
+- `.eq(value)`: Assign `value`.
+  - If `value` is a `Value`, it'll be assigned directly to the underlying `Value`.
+  - If `value` is an `int` or `float`, it'll be cast to a `fixed.Const` first.
+  - If `value` is a `fixed.Value`, the precision will be extended or truncated as required.
+- `.reshape(f_bits)`: Return a new `fixed.Value` with `f_bits` fractional bits, truncating or extending precision as required.
+- `.reshape(shape)`: Return a new `fixed.Value` with shape `shape`, truncating or extending precision as required.
```

Comment:
The 2 signatures of `.reshape()` here are my error. They should be collapsed into one `.reshape()`, although I am also considering dropping the second form if I can't find a killer use-case for it.

Reply:
In any case I think this method could use further bikeshedding. I would be just as open to calling this method `truncate(f_bits=0)` or `trunc(f_bits=0)`, although that loses the notion that it can both increase *and* decrease precision, which was part of the reason of not calling things `round()` in the first place.